Backend/app/rider-platform/rider-app/Main/src/API/UI/SearchRetry.hs

```diff
@@ -0,0 +1,154 @@
+import Kernel.Types.Id
+import Kernel.Types.SlidingWindowLimiter
+import Kernel.Utils.Common
+import Kernel.Utils.SlidingWindowLimiter
+import Servant hiding (throwError)
+import qualified SharedLogic.CallBPP as CallBPP
+import Storage.CachedQueries.CacheConfig
+import qualified Storage.CachedQueries.Merchant as QMerchant
+import qualified Storage.CachedQueries.RouteInfo as Redis
+import qualified Storage.Queries.Person as Person
+import Tools.Auth
+import qualified Tools.Maps as Maps
+import Tools.Metrics
+
+------ Search Flow --------
+
+type API =
+  "rideSearchRetry"
+    :> TokenAuth
+    :> ReqBody '[JSON] SearchRetryReq
+    :> Capture "parentSearchId" (Id SearchRequest)
+    :> Header "device" Text
+    :> Post '[JSON] SearchRetryRes
+
+handler :: FlowServer API
+handler = search
+
+data SearchRetryRes = SearchRetryRes
+  { searchId :: Id SearchRequest,
+    searchExpiry :: UTCTime,
+    routeInfo :: Maybe Maps.RouteInfo
+  }
+  deriving (Generic, FromJSON, ToJSON, Show, ToSchema)
+
+search :: Id Person.Person -> SearchRetryReq -> Id SearchRequest -> Maybe Text -> FlowHandler SearchRetryRes
+search personId req parentSearchId device = withFlowHandlerAPI . withPersonIdLogTag personId $ do
+  checkSearchRateLimit personId
+  (searchId, searchExpiry, routeInfo) <- searchRetry personId parentSearchId req device
+  return $ SearchRetryRes searchId searchExpiry routeInfo
+
+searchRetry ::
+  Id Person.Person ->
+  Id SearchRequest ->
+  DOneWayRetry.SearchRetryReq ->
+  Maybe Text ->
+  Flow (Id SearchRequest, UTCTime, Maybe Maps.RouteInfo)
+searchRetry personId parentSearchId req device = do
+  person <- Person.findById personId >>= fromMaybeM (PersonDoesNotExist personId.getId)
+  merchant <- QMerchant.findById person.merchantId >>= fromMaybeM (MerchantNotFound person.merchantId.getId)
+  dSearchRes <- DOneWayRetry.oneWaySearchRetry person merchant req parentSearchId
+  mbRouteResponse <- Redis.getRouteInfo parentSearchId
+  routeResponse <-
+    case mbRouteResponse of
+      Just a -> pure a
+      Nothing -> getrouteResponse dSearchRes person
+  let shortestRouteInfo = getRouteInfoWithShortestDuration routeResponse
+  let shortestRouteDistance = (.distance) =<< shortestRouteInfo
+  let shortestRouteDuration = (.duration) =<< shortestRouteInfo
+  fork "search cabs" . withShortRetry $ do
+    becknTaxiReq <- TaxiACL.buildOneWaySearchReq dSearchRes device shortestRouteDistance shortestRouteDuration (Just parentSearchId)
+    void $ CallBPP.search dSearchRes.gatewayUrl becknTaxiReq
+  return (dSearchRes.searchId, dSearchRes.searchRequestExpiry, shortestRouteInfo)
+
+checkSearchRateLimit ::
+  ( Redis.HedisFlow m r,
+    CoreMetrics m,
+    HasFlowEnv m r '["slackCfg" ::: SlackConfig],
+    HasFlowEnv m r '["searchRateLimitOptions" ::: APIRateLimitOptions],
+    HasFlowEnv m r '["searchLimitExceedNotificationTemplate" ::: Text],
+    MonadTime m
+  ) =>
+  Id Person.Person ->
+  m ()
+checkSearchRateLimit personId = do
```

Comment:
Many functions duplicated from `API.UI.Search`
Can we create some `SharedLogic.Search` module and put it there?

Reply:
done